### PR TITLE
Add pins to opam-install

### PR DESCRIPTION
**Pull request by davesnx**
While setting up the dev enviroment in opam, I found helpful to have the pins as part of the install.

Note: one annoyance is keeping CI in sync with the dune hash. Maybe running make commands in CI is a good idea? or having an env var?

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -24,6 +25,12 @@ opam-create-switch: ## Create opam switch
 opam-install: ## Install development dependencies
 	cd ocaml-tree && npm install
 	opam install -y ocaml-lsp-server
+	env IGNORECONSTRAINTS="melange,mel"
```
Comment (jchavarri) [inline Makefile:28]: Why is this needed?

File: Makefile [hunk 2]
```diff
@@ -24,6 +25,12 @@ opam-create-switch: ## Create opam switch
...
+	opam pin add dune https://github.com/ocaml/dune.git#0d44bbfdb2a68907a464aeb2dabe95388dac5712 -y
+	opam pin add melange-compiler-libs --dev-repo -y
```
Comment (jchavarri) [inline Makefile:30]: Not sure it is required to pin this one.

File: Makefile [hunk 3]
```diff
@@ -24,6 +25,12 @@ opam-create-switch: ## Create opam switch
...
+	opam install luv reason -y
+	opam pin add melange . --with-test -y
+	opam pin add mel . --with-test -y
```
Comment (jchavarri) [inline Makefile:33]: Same about these two (we just need to pin the dependencies of melange packages, not the melange packages themselves).

**Review by jchavarri — COMMENTED** [review]
I am not sure I get why pinning the melange packages (`melange-compiler-libs`, `melange` and `mel` is required.

**Comment by davesnx** [comment]
I'm not entirely sure why `-ignore-constraints` is needed for mel and melange and why we pin mel, melange and melange-compiler-libs. 

What I wanted to add here was a command to setup the environment the same as CI. Probably pinning dune and installing luv and reason was good enough, but I wasn't sure if there was something I could break.

**Comment by anmonteiro** [comment]
@jchavarri has been working in the melange codebase for a little while now. Perhaps we could record your dev setup script in the makefile for other opam user?

**Comment by jchavarri** [comment]
It is hard to know what I had done before, as opam is stateful and one pins stuff, unpins, etc.

So I just created a fresh opam switch and the only thing I needed to add to the `opam-install` command was the pinning of dune, and install `reason` (`luv` was already installed, as it's a dep of `mel`).

The commands I have tested are:
```
dune build @runtest -p melange
make dev
```

@davesnx Maybe we could leave the dune pin in the Makefile as is in this PR, and add `"reason" {with-test}` to `melange.opam`? That should do it, for dev setup.

**Comment by davesnx** [comment]
Closing in favor of https://github.com/melange-re/melange/pull/503
